jwst/datamodels/library.py

```diff
@@ -246,28 +246,51 @@ def get_crds_parameters(self):
                 )
                 idx = 0
 
-            # find model in _loaded_models, temp_filenames, or asn_dir
-            if self._on_disk:
-                if idx in self._temp_filenames:
-                    # if model has been modified, find its temp filename
-                    filename = self._temp_filenames[idx]
-                else:
-                    # otherwise, find the filename in the asn_dir
-                    member = self._members[idx]
-                    filename = Path(self._asn_dir) / member["expname"]
+            return self.read_metadata(idx)
+
+    def read_metadata(self, idx):
+        """
+        Read metadata for a model at the given index.
+
+        Parameters
+        ----------
+        idx : int
+            Index of the model in the library.
+
+        Returns
+        -------
+        dict
+            The metadata dictionary for the model.
+        """
+        if not self._open:
+            raise ClosedLibraryError("ModelLibrary is not open")
+
+        # if model was already borrowed, raise
+        if idx in self._ledger:
+            raise BorrowError("Attempt to read metadata from model that is already borrowed")
+
+        # find model in _loaded_models, temp_filenames, or asn_dir
+        if self._on_disk:
+            if idx in self._temp_filenames:
+                # if model has been modified, find its temp filename
+                filename = self._temp_filenames[idx]
             else:
-                if idx in self._loaded_models:
-                    # if this model is in memory, retrieve parameters from it directly
-                    model = self._loaded_models[idx]
-                    return model.get_crds_parameters()
-                else:
-                    # otherwise, find the filename in the asn_dir
-                    member = self._members[idx]
-                    filename = Path(self._asn_dir) / member["expname"]
-
-            meta = read_metadata(filename, flatten=True)
-
-        return meta
+                # otherwise, find the filename in the asn_dir
+                member = self._members[idx]
+                filename = Path(self._asn_dir) / member["expname"]
+        else:
+            if idx in self._loaded_models:
+                # if this model is in memory, retrieve parameters from it directly
+                # the desired behavior is already implemented by get_crds_parameters
+                # so we can just call it here, it is not inherently CRDS-specific
+                model = self._loaded_models[idx]
+                return model.get_crds_parameters()
+            else:
+                # otherwise, find the filename in the asn_dir
+                member = self._members[idx]
+                filename = Path(self._asn_dir) / member["expname"]
+
+        return read_metadata(filename)
```

Comment:
Does this handle assigning metadata from the association to the model metadata (like[_assign_member_to_model](https://github.com/spacetelescope/stpipe/blob/e114f3bdff075d2c42c0bed4fac284e72af2976a/src/stpipe/library.py#L506) that is used during borrow)?

Reply:
No it doesn't.  That's a good suggestion though, for consistency with what happens to those metadata when you borrow a model.  I will update this

Reply:
how does https://github.com/spacetelescope/jwst/pull/9570/commits/48f0bf246df735c2945769d307c92e1444e5ec40 look?